src/general/world_effects/patch_events/MeteorImpactEvent.cs

```diff
@@ -0,0 +1,308 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Godot;
+using Newtonsoft.Json;
+using Xoshiro.PRNG64;
+
+[JSONDynamicTypeAllowed]
+public class MeteorImpactEvent : IWorldEffect
+{
+    private const string TemplateBiomeForChunks = "patch_event_template_biome";
+
+    [JsonProperty]
+    private readonly HashSet<int> modifiedPatchesIds = new();
+
+    [JsonProperty]
+    private readonly XoShiRo256starstar random;
+
+    [JsonProperty]
+    private GameWorld targetWorld;
+
+    [JsonProperty]
+    private Meteor selectedMeteor = null!;
+
+    public MeteorImpactEvent(GameWorld targetWorld, long randomSeed)
+    {
+        this.targetWorld = targetWorld;
+        random = new XoShiRo256starstar(randomSeed);
+    }
+
+    [JsonConstructor]
+    public MeteorImpactEvent(GameWorld targetWorld, XoShiRo256starstar random)
+    {
+        this.targetWorld = targetWorld;
+        this.random = random;
+    }
+
+    public void OnRegisterToWorld()
+    {
+    }
+
+    public void OnTimePassed(double elapsed, double totalTimePassed)
+    {
+        FinishEvent();
+        TryToTriggerEvent(totalTimePassed);
+    }
+
+    private bool IsSurfacePatch(Patch patch)
+    {
+        return patch.Depth[0] == 0 && patch.BiomeType != BiomeType.Cave;
+    }
+
+    private void TryToTriggerEvent(double totalTimePassed)
+    {
+        if (!AreConditionsMet())
+            return;
+
+        ChooseAffectedPatches();
+        ChooseMeteorType();
+
+        foreach (var patch in targetWorld.Map.Patches.Values)
+        {
+            if (modifiedPatchesIds.Contains(patch.ID))
+            {
+                ChangePatchProperties(patch, totalTimePassed);
+            }
+        }
+
+        LogBeginningOfMeteorStrike();
+    }
+
+    private void ChooseMeteorType()
+    {
+        var meteors = SimulationParameters.Instance.GetAllMeteors().ToList();
+        var index = GetRandomElementByProbability(meteors.Select(meteor => meteor.Probability).ToList(),
+            random.NextDouble());
+        selectedMeteor = meteors.ElementAt(index);
+    }
+
+    private int GetRandomElementByProbability(List<double> chances, double probability)
+    {
+        double totalSum = chances.Sum();
+        if (Math.Abs(totalSum - 1.0) > 1e-6)
+        {
+            GD.PrintErr($"Probability sum mismatch: {totalSum}. Expected: 1.0");
+            return 0;
+        }
```

Comment:
Isn't this static info? This should be checked by SimulationParameters when loading, not during gameplay.